### PR TITLE
Fix constraint generation properly

### PR DESCRIPTION
**Pull request by potiuk**
The #17939 did not fix the problem finally. It turned out that
one more change was needed - since we now always upgrade to latest
dependencies in `push` and `schedule` type of build we do not need
to check for the variable UPGRADE_TO_NEWER_DEPENDENCIES (which
was not set in "Build Image" step.

This fixes it, but also changes the constraint generation to add
comments in the generated constraint files, describing how and
why the files are generated.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by uranusjr** [comment]
Is this supposed to fix the CI image generation on this PR, or will it only take effect after merged into main?

File: scripts/ci/selective_ci_checks.sh [hunk 1]
```diff
@@ -353,8 +351,9 @@ function check_if_setup_files_changed() {
 
     if [[ $(count_changed_files) != "0" ]]; then
         # In case the setup files changed, we automatically force upgrading to newer dependencies
-        # no matter what was set before
-        upgrade_to_newer_dependencies="${INCOMING_COMMIT_SHA}"
+        # no matter what was set before. We set it to random number to make sure that it will be
+        # always invalidating the layer in Docker that triggers installing the dependencies
+        upgrade_to_newer_dependencies="${RANDOM}"
```
Comment (uranusjr) [inline scripts/ci/selective_ci_checks.sh:356]: Why is this needed? (Also the one in the previous block.)

**Review by ashb — APPROVED** [review]

File: scripts/ci/selective_ci_checks.sh (hunk 1 above)
Comment (potiuk) [reply scripts/ci/selective_ci_checks.sh:356]: Asking for RANDOM ?

The reason is that we need to make sure that the right layer is invalidated when the Dockerfile is rebuilt in "eager upgrade" even if the `setup.py` does not change (when we build it in `main`). 

I just updated the description (see the latest version) where it was described in detail (see the changes in CI.rst). 

I used `commit SHA` before, but the problem is that we have different commits in `main` and `pr` kind of builds (in Github actions they are passed in different variables in this case). But RANDOM will do the same thing. What we want here we want to invalidate everything in the image from that place where ARG UPGRADE_TO_NEWER_DEPENDENCIES is used.

Everything below that line should be invalidated - regardless from `setup.py`.

https://github.com/apache/airflow/blob/7c4d3173f2c73ceeadb7155cf367983a90c6dca8/Dockerfile.ci#L282

It could `technically` be just `true` (and it was initially actually), but having RANDOM here makes it more "certain" that it will be invalidated in case we build the image with a cache.

**Comment by kaxil** [comment]
🤞

**Comment by rickiesmooth** [comment]
Did this PR change the output of `https://raw.githubusercontent.com/apache/airflow/constraints-1.10.12/constraints-3.7.txt`? I use this file in my `requirements.txt` as suggested by [MWAA](https://docs.aws.amazon.com/mwaa/latest/userguide/working-dags-dependencies.html) and it suddenly started to fail installing because the dependencies didn't match anymore. After updating my dependencies everything worked again so just wanted to let you know about that!

**Comment by ashb** [comment]
@rickiesmooth Nope, those constraint files haven changed in 14 months https://github.com/apache/airflow/tree/constraints-1.10.2

**Comment by rickiesmooth** [comment]
ah ok thanks for checking!
